tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala

```diff
@@ -226,16 +232,277 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
       emitNextLine(0)
   }
 
+  // Private API for Rhino
+
+  private[scalajs] object rhinoAPI { // scalastyle:ignore
+    private implicit val GlobalKnowledge: GlobalKnowledge = new Invalidatable {
+      override def invalidate(): Unit = ()
+    }
+
+    def initialize(linkingUnit: LinkingUnit): Unit =
+      startRun(linkingUnit)
+
+    def getHeaderFile(): org.scalajs.core.tools.io.VirtualJSFile =
+      CoreJSLibs.lib(semantics, outputMode)
+
+    def genClassDef(linkedClass: LinkedClass): js.Tree =
+      classEmitter.genClassDef(linkedClass)
+  }
+
+  // Guardian of all global knowledge
+
+  private abstract class InvalidatableBag {
+    def forgetAbout(invalidatable: Invalidatable): Unit
+  }
+
+  private object KnowledgeGuardian { knowledgeGuardian =>
```

Comment:
> The knowledge guardian is logically coupled to maintaining the caches.

Are you sure about that? It seems that the point of the knowledge guardian precisely is to decouple caching and invalidation logic. The knowledge guardian does not know about caches (neither it should). All it does is track what was using which assumptions and invalidate these elements as needed.

The Emitter on the other hand ties caches and Invalidatables together. Think of it that way: It would be absolutely sensible to use the knowledge guardian to implement a piece of code that replaces code on the fly in a VM every new linker run. There are no caches (at least not in the Emitters sense) in that scenario, nevertheless, an instance of the knowledge guardian can be used unchanged.